packages/yew/src/html/component/lifecycle.rs

```diff
@@ -318,43 +322,57 @@ impl Runnable for PropsUpdateRunner {
-                    next_sibling: ref mut current_next_sibling,
-                    ..
-                } => {
-                    // When components are updated, their siblings were likely also updated
-                    *current_next_sibling = next_sibling;
-                    // Only trigger changed if props were changed
-                    state.inner.props_changed(props)
-                }
-
-                #[cfg(feature = "hydration")]
-                ComponentRenderState::Hydration {
-                    next_sibling: ref mut current_next_sibling,
-                    ..
-                } => {
-                    // When components are updated, their siblings were likely also updated
-                    *current_next_sibling = next_sibling;
-                    // Only trigger changed if props were changed
-                    state.inner.props_changed(props)
+            if let Some(next_sibling) = next_sibling {
+                // When components are updated, their siblings were likely also updated
+                // We also need to shift the bundle so next sibling will be synced to child
+                // components.
+                match state.render_state {
+                    #[cfg(feature = "csr")]
+                    ComponentRenderState::Render {
+                        next_sibling: ref mut current_next_sibling,
+                        ref parent,
+                        ref bundle,
+                        ..
+                    } => {
+                        bundle.shift(parent, next_sibling.clone());
+                        *current_next_sibling = next_sibling;
+                    }
+
+                    #[cfg(feature = "hydration")]
+                    ComponentRenderState::Hydration {
+                        next_sibling: ref mut current_next_sibling,
+                        ref parent,
+                        ref fragment,
+                        ..
+                    } => {
+                        fragment.shift(parent, next_sibling.clone());
+                        *current_next_sibling = next_sibling;
+                    }
+
+                    #[cfg(feature = "ssr")]
+                    ComponentRenderState::Ssr { .. } => {
+                        #[cfg(debug_assertions)]
+                        panic!("properties do not change during SSR");
+                    }
                 }
+            }
 
-                #[cfg(feature = "ssr")]
-                ComponentRenderState::Ssr { .. } => {
-                    #[cfg(debug_assertions)]
-                    panic!("properties do not change during SSR");
-
-                    #[cfg(not(debug_assertions))]
+            // Only trigger changed if props were changed / next sibling has changed.
+            let schedule_render = match state.has_rendered {
+                true => state.inner.props_changed(props),
+                false => {
+                    state.pending_props = Some(props);
                     false
```

Comment:
So this sets `schedule_render = false` but the `pending_props` are only taken `if schedule_render`. That seems to mess up the order in which properties arrive:
1. prop update while `!state.has_rendered` --> `state.pending_props = Some(_)`, but because `!schedule_render`, no `PropsUpdateRunner` is scheduled.
2. component renders normally, setting `state.has_rendered = true`
3. another prop update, with `state.inner.props_changed(props) == true` --> `schedule_render == true` and a `PropsUpdateRunner` **with the old, outdated** props from `state.pending_props.take()` is run, overriding the newly set ones?!

Can you clarify?

Reply:
This is safe as has_rendered is only flipped at render cycle which is run 1 at a time and at that time any pending props update are run already (as props_update can only be scheduled externally from parent render lifecycle and parent always render before children).

But let's switch to a more robust way of doing it in case this changes in the future.